src/main/scala/io/iohk/ethereum/ledger/BlockQueue.scala

```diff
@@ -0,0 +1,187 @@
+package io.iohk.ethereum.ledger
+
+import akka.util.ByteString
+import io.iohk.ethereum.domain.{Block, Blockchain}
+import io.iohk.ethereum.ledger.BlockQueue.{Leaf, QueuedBlock}
+import io.iohk.ethereum.utils.Config.SyncConfig
+import io.iohk.ethereum.utils.Logger
+import org.spongycastle.util.encoders.Hex
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+
+object BlockQueue {
+  case class QueuedBlock(block: Block, totalDifficulty: Option[BigInt])
+  case class Leaf(hash: ByteString, totalDifficulty: BigInt)
+}
+
+class BlockQueue(blockchain: Blockchain, syncConfig: SyncConfig) extends Logger {
+
+  private val blocks = mutable.Map[ByteString, QueuedBlock]()
```

Comment:
what do you think about making it synchronized in case someone invokes Ledger from multiple threads/actors?

Reply:
`synchronized` sounds good as I believe actorising this would be an overkill. But do you think we should do it now?

Would there be any practical downsides of having it synchronized?

Even if do we it, the `Ledger` would still not be thread-safe, we would not to also synchronize the `Blockchain`, right?

Reply:
as I'm thinking now, maybe it would be enough to add comments on these Ledger methods that they should only be called from one thread (i.e RegularSync). wdyt?